### PR TITLE
Move weight computation from propose to traced_evaluate.

### PR DESCRIPTION
**Pull request by zmheiko**

### CODE REVIEW

File: coix/numpyro.py [hunk 1]
```diff
@@ -42,11 +45,18 @@ def wrapped(*args, **kwargs):
         trace[name] = {"value": value, "log_prob": log_prob}
         if site.get("is_observed", False):
           trace[name]["is_observed"] = True
+    # else:
+    # in_log_weight = _get_log_weight(q_trace, batch_ndims)
```
Comment (fehiepsi) [inline coix/numpyro.py:49]: this can be deleted I think

File: coix/numpyro.py [hunk 2]
```diff
@@ -42,11 +45,18 @@ def wrapped(*args, **kwargs):
...
     metrics = {
         name: site["value"]
         for name, site in tr.items()
         if site["type"] == "metric"
     }
+    # add log_weight to metrics
+    if "log_weight" not in metrics:
+      log_probs = [get_site_log_prob(site) for site in trace.values()]
+      weight = _get_log_weight(trace, _get_batch_ndims(log_probs))
+      metrics = {**metrics, "log_weight": weight}
```
Comment (fehiepsi) [inline coix/numpyro.py:59]: after this, I think we can add a note to `compose` that currently, if one of q1 or q2 is an inference program, then the other is a primitive without observed variables (a kernel?).

**Review by fehiepsi — COMMENTED** [review]
Thanks Heiko! I'll take another pass to the tutorials over the weekends.

File: coix/algo.py [hunk 3]
```diff
@@ -215,7 +215,7 @@ def body_fun(i, q):
       p, q = extend(targets(i + 1), reverses(i)), compose(forwards(i), q)
       return propose(p, q, loss_fn=avo_loss, detach=True)
 
-    return fori_loop(0, num_targets - 1, body_fun, targets(0))
+    return fori_loop(0, num_targets - 1, body_fun, body_fun(0, targets(0)))
```
[the remark this answers is not in the document]
Comment (zmheiko) [reply coix/algo.py:218]: You are correct, it should be `fori_loop(1, ...)`. I made the change a while ago because I had an issue with the structure of the initial loop argument when running AVO. Looking at it again it seems that the change is indeed not needed. The `propose` returned by the loop body returns the return value of `targets(i + 1)` when executed, which should have the same structure as the return value of `targets(0)` (in most use cases).

[the remark this answers is not in the document]
Comment (zmheiko) [reply coix/algo.py:218]: Oh I see. I had another local change that made `extend` return the concatenated return value of both programs, which is probably why I thought the change is necessary. So the change is definitely not needed!

File: coix/numpyro.py (hunk 2 above)
Comment (zmheiko) [reply coix/numpyro.py:59]: Yes, I call them kernel programs in the tutorials now. I think that's an appropriate name.

**Comment by zmheiko** [comment]
Also regarding the notebooks. A wanted to make that a new PR but accidentally pushed the changes to the same branch. Would it be better to still remove them and make a new PR?

**Comment by fehiepsi** [comment]
Yes, it's better to put it in a separate PR.

File: coix/algo.py [hunk 4]
```diff
@@ -289,7 +289,7 @@ def body_fun(i, q):
       p, q = extend(p, rev), compose(stick_the_landing(fwd), resample(q))
       return propose(p, q, loss_fn=rkl_loss, detach=True)
 
-    return fori_loop(0, num_targets - 1, body_fun, targets(0))
+    return fori_loop(0, num_targets - 1, body_fun, body_fun(0, targets(0)))
```
Comment (fehiepsi) [inline coix/algo.py:292]: I guess we can also revert this change?

**Review by fehiepsi — APPROVED** [review]
